k8srm-prototype/pkg/api/capacity_types.go

```diff
@@ -38,19 +39,91 @@ type DevicePoolSpec struct {
 	// Attributes contains device attributes that are common to all devices
 	// in the pool.
 	// +optional
+	// +listType=atomic
 	Attributes []Attribute `json:"attributes,omitempty"`
 
-	// DeviceCount contains the total number of devices in the pool.
+	// SharedResources are pooled resources that are shared by all devices
+	// in the pool. This is typically used when representing a
+	// partitionable device, and need not be populated otherwise.
+	//
+	// +optional
+	// +listType=atomic
+	SharedResources []ResourceCapacity `json:"sharedResources,omitempty"`
+
+	// Devices contains the individual devices in the pool.
+	//
 	// +required
-	DeviceCount int `json:"count,omitempty"`
+	// +listType=atomic
+	Devices []Device `json:"devices,omitempty"`
 }
 
 // DevicePoolStatus contains the state of the pool as last reported by the
 // driver. Note that this will not include the allocations that have been made
 // by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
 // to make future scheduling decisions.
 type DevicePoolStatus struct {
-	AvailableDevices int `json:"availableDevices,omitempty"`
+	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
+}
+
+// AllocatedDevice represents a device that has been allocated from the pool.
+type AllocatedDevice struct {
+	Name string
+	// Conditions contains the latest observation of the device's state.
+	Conditions []metav1.Condition `json:"conditions"`
+
+	// ClaimUIDs contains the UIDs of the claims to which this device
+	// is allocated.
+	ClaimUIDs []types.UID
```

Comment:
We should comment why this makes sense as a plural

Reply:
I pluralized this when I made it a list. If we allowed devices to be shared between claims (ie, partial device usage), then we need this. That is a current use case: sharing mediated by the platform, not the user.

Sharing mediated by the user is done by pointing multiple pods at the same claim. We have discussed that at some length and it seems to be a use case for NVIDIA at least.

But suppose you have independent users and you want to be able to share devices between them. This is exactly the model today for nodes. We also do it today with Device Plugin by having the plugin advertise the same care multiple times. We would like to allow that, but do it in a more controlled way. This is where per-device allocatable resources come in. I can say "I need a GPU with 8GB". We have a share-able GPU with 32GB, so we give you access to that. The next user comes in and says, "I need a GPU with 16GB". We know you will fit, so we can put you on the same GPU. At that point, two independent claims land on the same underlying device.

The DeviceClaimStatus will show the allocations and status from the point of view of the individual claims. The DevicePoolStatus shows it from the device's point of view. This field allows us to track which claims are using this device.